### PR TITLE
allow better grpc error messages from Go component providers

### PR DESCRIPTION
**Pull request by tgummerer**
The Go version of providing better error details.

Compared to Python and NodeJS I ended up deciding only to add a single helper method, since it works out quite nicely in Go with varargs.  Could definitely be convinced to add a `NewInputPropertyError` function though.

### CODE REVIEW

File: sdk/go/pulumi/errors.go [hunk 1]
```diff
@@ -0,0 +1,40 @@
+// Copyright 2016-2024, Pulumi Corporation.
```
Comment (lunaris) [inline sdk/go/pulumi/errors.go:1]:
```suggestion
// Copyright 2024, Pulumi Corporation.
```

File: sdk/go/pulumi/errors.go [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+	for _, e := range errors {
+		errorDetails.Errors = append(errorDetails.Errors, &pulumirpc.InputPropertiesError_PropertyError{
+			PropertyPath: e.PropertyPath,
+			Reason:       e.Reason,
+		})
+	}
+
+	s, _ := status.Newf(codes.InvalidArgument, "%s", message).WithDetails(&errorDetails)
+	return s.Err()
+}
```
Comment (lunaris) [inline sdk/go/pulumi/errors.go:40]: I don't mind how we land on this, but I feel like we should not abuse the fact that we have the native gRPC types available to us here. I'd support us defining our own error types e.g. 

```go
type InputPropertiesError struct {
  Err error
  Details []InputPropertyErrorDetails
}
```

and looking for these in the server and translating appropriately. I don't think it'll cost us hugely not to separate concerns like this, but e.g. if we do I think it makes some things neater -- e.g. customers who want to test component providers or the like can assert that `errors.As(..., pulumi.InputPropertiesError, ...)` or the like without having to get down and dirty with gRPC.

File: sdk/go/pulumi/errors.go [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+	// Convert the errors to a slice of proto messages.
+	errorDetails := pulumirpc.InputPropertiesError{}
+	for _, e := range errors {
+		errorDetails.Errors = append(errorDetails.Errors, &pulumirpc.InputPropertiesError_PropertyError{
+			PropertyPath: e.PropertyPath,
+			Reason:       e.Reason,
+		})
+	}
+
+	s, _ := status.Newf(codes.InvalidArgument, "%s", message).WithDetails(&errorDetails)
```
Comment (lunaris) [inline sdk/go/pulumi/errors.go:38]: Related to the comment below, I feel like a fluent-style pattern has more appeal in Go due to the way errors work/the prevalence of `Errorf`-like helpers, etc. Maybe something like:

```go
type InputPropertiesError struct { ... }
type InputPropertyErrorDetails struct { ... }

func NewInputPropertyError(propertyPath string, reason string) *InputPropertiesError
func InputPropertyErrorf(propertyPath string, format string, args ...interface{}) *InputPropertiesError
func NewInputPropertiesError(message string, details ...InputPropertyErrorDetails) *InputPropertiesError
func InputPropertiesErrorf(format string, args ...interface{}) *InputPropertiesError

func (ipe *InputPropertiesError) Error() string
func (ipe *InputPropertiesError) Unwrap() string
func (ipe *InputPropertiesError) WithDetails(details ...InputPropertyErrorDetails) *InputPropertiesError
```

could work? But again happy to be shot down on this!

File: sdk/go/pulumi/errors.go (hunk 1 above)
Comment (lunaris) [reply sdk/go/pulumi/errors.go:1]: (same elsewhere)

File: tests/integration/construct_component_failures/go/main.go [hunk 4]
```diff
@@ -0,0 +1,21 @@
+// Copyright 2016-2024, Pulumi Corporation.  All rights reserved.
```
Comment (lunaris) [inline tests/integration/construct_component_failures/go/main.go:1]: If this passes the `goheader` lint, I guess we don't lint test Go code?

File: sdk/go/pulumi/errors.go (hunk 2 above)
Comment (tgummerer) [reply sdk/go/pulumi/errors.go:40]: Yeah that makes sense.  I'm happy to have a separate error type here.  I would make it 

```
type InputPropertiesError struct {
  Message string
  Errors []InputPropertyErrorDetails
}
```

though for consistency with other languages :thinking: I think that also matches the interface you're proposing below.

File: tests/integration/construct_component_failures/go/main.go (hunk 4 above)
Comment (tgummerer) [reply tests/integration/construct_component_failures/go/main.go]: Apparently not :)  I made it consistent with the other go files now.

File: sdk/go/pulumi/errors.go (hunk 3 above)
Comment (tgummerer) [reply sdk/go/pulumi/errors.go]: This works for me!

File: sdk/go/pulumi/errors/errors.go [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+	if message != "" && len(ipe.Errors) > 0 {
+		message += ": "
+	}
+	for i, err := range ipe.Errors {
+		if i == 0 {
+			message += ": "
+		}
+		message += fmt.Sprintf("%s: %s", err.PropertyPath, err.Reason)
+	}
+	return message
```
Comment (lunaris) [inline sdk/go/pulumi/errors/errors.go:65]: This logic looks a little shaky to me -- feels like we might end up with too many colons? And perhaps we have missed a newline?

I'd be tempted TBH to give `InputPropertyErrorDetails` a `String()` which does `<propertyPath>: <reason>` then you can just defer to that in the loop here, but happy either way.

File: sdk/go/pulumi/errors/errors.go [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package errors
+
+import "fmt"
+
+type InputPropertyErrorDetails struct {
```
Comment (lunaris) [inline sdk/go/pulumi/errors/errors.go:19]: I think it'd be good to doc-comment the things we're adding to the public API, though equally happy to wait on that if we expect a bit more discussion on what the API should look like.

**Review by lunaris — APPROVED** [review]

File: sdk/go/pulumi/errors/errors.go (hunk 5 above)
Comment (tgummerer) [reply sdk/go/pulumi/errors/errors.go:65]: I wasn't sure about having newlines in error strings tbh.  But maybe that's fine and we should just do it.  `InputPropertyErrorDetails` having a `String()` method sounds like a good idea either way!

**Review by justinvp — APPROVED** [review]

**Pull request merged**